### ISSUE TITLE
Component within a partial calling onint() and onrender() when already renderred

### ISSUE THREAD

**Issue by krityan** [ractivejs/ractive#3151]
## Description:

We have a dynamically generated partial that contains a component, when we upgraded to 0.9.10 we noticed that this component has it's oninit()/onrender() functions called multiple times, rather than only being called when it is rendered like before.

Tracing it back, this first became an issue in 0.9.6 as it works fine in 0.9.5, however in 0.9.6 it caused the browser to hang. In 0.9.7 it instead causes the component to re-render and trigger its oninit and onrender calls. Since these are supposed to be for initialising the component this causes the component to reinitialise whenever this re-render is triggered. This seems to be triggered by us resetting a partial within the component.

Notes about when it happens:

Action that leads to issue: updating properties of an object property called model (via .set('model.value', xxx).  
In the component that re-initialises, model.xxx is one of the dependencies for a computed property called tileDefinition.  
Model has an observer in the same component, as does tileDefinition.  
After both observers fire is when the component re-initialises.  
Have made both these observers do nothing, re-initialisation still happens so content of observers isn't the issue.  
One of the partials in this component updates at the same time, and seems to be the trigger. It renders a component, that component doesn't change but the content of it does.

We're not sure if this is intentional, since we're not doing anything to make the component unrender/render again so this isn't the expected behaviour for us.

If this is intentional, do you have any advice for how to work around it?

## Versions affected:

0.9.5 - fine 
0.9.6 - browser hangs
0.9.7 -> 0.9.10 - component re-renders causing re-initialisation

## Platforms affected:

Chrome 62 / IE 11

## Reproduction:

Create a component within a partial, when a partial of that component is reset it causes the whole component to rerender, triggering oninit and onrender.

**Comment by ghost** [comment]
Hi! Can you provide a repro in [playground](https://ractive.js.org/playground/)? Minimal/non-working is fine (stuff tends to work on minimal) as long as we get the general idea of your current approach.

**Comment by krityan** [comment]
In the process of doing so, will get back to you when it's finished

**Comment by krityan** [comment]
This is a very basic idea of what our tiles are doing: [playground](https://ractive.js.org/playground/#N4IgFiBcoE5SAbAhgFwKYGcUgL4BoQN4AdAOwCUkBjFASwDc0A6KgewFsAHV0tUlDEwAmtDJ2QBPAMIduvfgAIAvAso0GzNAA90pIQApgZYinRdk6SAoDkAYhFjJMrjz4preYyi9DUSKwBmAK6k6jz6AJQKRqQmJjBoKEEwpNFecSjsrEJoCFYxGRlYQQEBVsQgFeneKDjVdbG1no1ePLSktCiBIWGkkWktpmCiTOHWaCIoALLZuR4KwaF04VEFNSYow4IYifrWWTkITMWl89akrAoIPADmaDAKm2gKOQFIQQjuEQDc9RFeDRw-1iFGodEYLFkrn4gjYCWccjcylUYI0TG0ugMazM4lQaCsdjhaAR0PczTijV8KH8Cx6yz6q2qCSSKQGhXQk3aN0CSAQO3J60yszybMFJhOZRsr3en2s1RMDQyOF+jXw6UanCQMDovIw+WqdAQaAAClqdSLrHLVQLWh0Ol1aUtaDwFP01hstqM+tYUKwbjcjQBRSbzRa9N3yoYjX3+o17CadLnWH5-FWFQabEZjHYoU3a2i80N050M0UZTOCBI5vPmvaGk1mgsIM4AHgcuOkUPkKAUB1ySmAwD7CBwOAUAHoAHzJtM1IGzwEpkFkehahQwVGMZG8ADuKPUjEMXmpMDuDrsJ7PHjICkeaHMeIJthxFjQ19SChvCjPNab3SdKwDLeJirg8mrasi1gtkSJLdhO04qsBUaCOB5oYAA2qhAC6yKoYhn4oMyySpHhX4NMBpBAiABDYJAIBtgwCi0EISgVJeiQVJOLbjiI9CTsYpAthgVAwLQnA9sxrEgC+eIVI8EicGgUnoDo44bgeaCcV4g6Tj+ja8pEo5kNxwmieJ-EgkJIliRJLEVDBXZuFpjQMXx1QtpwFnsgmdCkNy0TAD5XJGYM3GeV43G8V53guVF7kAEZBKYLo8AAtFQCC0FQADWyl+gGaDBp0nExgVCg+dxiXJaQ0WRQw0URXFgw6fWv68iFJh1W5LSxfVCVJb6qRpRlWW5RU1b6QgnE5goqFNpVA08LVPF9T1JnWeZAlWWZtlSe2TiOfwzmdU1GQeQ1NSDsOxxJKUHU1GFy2nTFJ2rWdVWDQow2ZTlUk+TMhycT5vbCgt1VPat607ZOuBAA)

Unfortunately I don't have the time to implement all of the data parts of it since they're quite complex, and I can't implement just the part that breaks it because isolating it is my current problem. When I've isolated it I'll implement it in the playground.

**Comment by krityan** [comment]
> Experimental parser transforms have been removed and replaced with a runtime construct associated with partials (see below)

Reading this, it seems that nothing has changed before the partial transforms, just now partials transforming is triggering on* life-cycle events? Unless I've misunderstood it.

**Comment by evs-chris** [comment]
Partials should only ever affect components downstream from (contained by) the partial. They don't trigger any lifecycle events directly, but resetting a partial that contains a component will tear down and recreate that component. You can try adding a console log statement during `teardown`, `init`, `render`, and `unrender` to see if that's what's happening.

**Comment by krityan** [comment]
Added some logs to our listeners and to the tiles oninit -> onteardown functions.

Only difference between these builds is the version of Ractive used.

`<core-tile>` is contained by a partial which we generate from a function call as seen in the playground example above.

In 0.9.6 the browser hangs when the user clicks on the new tile type.

```
0.9.10 (happens on 0.9.7)

<INITIAL LOAD>
[core-tile] oninit()
[core-tile] _tileDefinition updated
[core-tile] onrender()
[core-tile] _tileDefinition updated

<USER CLICKS ON NEW TILE TYPE>
[config-tile-selection] changeTile called
[core-tile] _tileDefinition updated

<RE-INITIALISATION>
[core-tile] onunrender()
[core-tile] onteardown()
[core-tile] oninit()
[core-tile] _tileDefinition updated
[core-tile] onrender()
[core-tile] model updated

0.9.3 (happens on 0.9.5)

<INITIAL LOAD>
[core-tile] oninit()
[core-tile] _tileDefinition updated
[core-tile] onrender()
[core-tile] _tileDefinition updated

<USER CLICKS ON NEW TILE TYPE>
[config-tile-selection] changeTile called
[core-tile] _tileDefinition updated
[core-tile] _tileDefinition updated
[core-tile] model updated
```

**Comment by evs-chris** [comment]
The only thing I could come up with was that the expression for the partial was being invalidated for some reason, which would reset the partial template. I added a strict equality check to the partial reset, so if it evals to the exact same thing, then no reset will occur. Please reopen if this doesn't fix the issue.

**Comment by krityan** [comment]
Just tried it with 1.0.0 build 110 and the teardown-render cycle is still present.

**Comment by evs-chris** [comment]
I'm not sure where else to look. Can you add a `console.log(new Error().stack)` to your construct, init, and teardown listeners? I may be able to figure out something from there.

Unless you have a full reproduction now?

**Comment by krityan** [comment]
Sorry I've been busy working on something else, below are the stack traces you wanted.

```
<INITIAL PAGE LOAD>

Error

at Child.onconstruct (Scripts)
at Hook.fire (Ractive)
at construct (Ractive)
at new Component (Ractive)
at createItem (Ractive)
at Fragment.createItems (Ractive)
at new Fragment (Ractive)
at createFragment$1 (Ractive)
at Partial.bind (Ractive)
at bind (Ractive)

Error

at Child.oninit (Scripts)
at Hook.fire (Ractive)
at fire (Ractive)
at fire (Ractive)
at fire (Ractive)
at fire (Ractive)
at fire (Ractive)
at fire (Ractive)
at fire (Ractive)
at HookQueue.end (Ractive)


<SELECT TILE TYPE>

Error

at Child.onteardown (Scripts)
at Hook.fire (Ractive)
at teardown$1 (Ractive)
at Component.unbind$1 [as unbind] (Ractive)
at unbind (Ractive)
at Array.forEach (<anonymous>)
at Fragment.unbind$4 [as unbind] (Ractive)
at Fragment.resetTemplate (Ractive)
at Partial.update (Ractive)
at update (Ractive)

Error

at Child.onconstruct (Scripts)
at Hook.fire (Ractive)
at construct (Ractive)
at new Component (Ractive)
at createItem (Ractive)
at Fragment.createItems (Ractive)
at Fragment.resetTemplate (Ractive)
at Partial.update (Ractive)
at update (Ractive)
at Array.forEach (<anonymous>)

Error

at Child.oninit (Scripts)
at Hook.fire (Ractive)
at fire (Ractive)
at HookQueue.end (Ractive)
at initialise (Ractive)
at Component.bind$2 [as bind] (Ractive)
at bind (Ractive)
at Array.forEach (<anonymous>)
at Fragment.bind$4 [as bind] (Ractive)
at Fragment.resetTemplate (Ractive)
```

**Comment by evs-chris** [comment]
Does the function that sets up your partial always return a unique name by any chance? If it does, then making sure that the partial name returned is the same on subsequent calls will resolve the issue. If not, can you throw a console log out with the partial name from the partial-making function?

**Comment by krityan** [comment]
Have checked the name of the partial being created at page load and when the tile type is changed in both 0.9.3 and 0.9.10, output is below.

To confirm, these are the names of the partials, we store them with the same name as the generated string for the partial itself.

```
0.9.10:

<page load>
Partial dynamically created: <core-tile  renderCore="{{true}}" model="{{this}}" context="{{context}}" state="{{state}}" listIndex="{{@index}}" />
<change tile>
Partial dynamically created: <core-tile  renderCore="{{true}}" model="{{this}}" context="{{context}}" state="{{state}}" listIndex="{{@index}}" />

0.9.3:

<page load>
Partial dynamically created: <core-tile  renderCore="{{true}}" model="{{this}}" context="{{context}}" state="{{state}}" listIndex="{{@index}}" />
<change tile>
Partial dynamically created: <core-tile  renderCore="{{true}}" model="{{this}}" context="{{context}}" state="{{state}}" listIndex="{{@index}}" />
```

**Comment by evs-chris** [comment]
Taking another swing at this, in your partial generation function are you doing something like this?

```js
function getPartial() {
  var name = 'partial here';
  this.partials[name] = name;
  return name;
}
```

If so, can you add in a check to see if the partial has already been created? I think the issue is that you're triggering a reparse on each invalidation because I'm guessing that there's a changing value referenced in your partial expression. If you add in something like this:

```js
function getPartial() {
  var name = 'partial here';
  if (name in this.partials) return name;
  this.partials[name] = name;
  return name;
}
```

it will stop causing the partial to invalidate fully.

**Comment by krityan** [comment]
That's exactly what we are doing, I've added the check and it's no longer re-adding the partial to this.partials but it hasn't resolved the issue with the component.

Something that might be worth mentioning is that there's actually 2 functions, one is called from html and returns a partial calling the other one which then returns the component code.

I added logs to both of these and can see that the core-tile partial isn't being re-added to this.partials, I can see that partials within core-tile are being readded when this happens but those shouldn't cause the whole component to teardown and reconstruct.
